crates/starknet-devnet-server/src/api/json_rpc/endpoints.rs

```diff
@@ -24,21 +24,33 @@ impl JsonRpcHandler {
 
     /// starknet_getBlockWithTxHashes
     pub async fn get_block_with_tx_hashes(&self, block_id: BlockId) -> StrictRpcResult {
-        let block =
-            self.api.starknet.read().await.get_block(block_id.as_ref()).map_err(
-                |err| match err {
-                    Error::NoBlock => ApiError::BlockNotFound,
-                    unknown_error => ApiError::StarknetDevnetError(unknown_error),
-                },
-            )?;
+        let starknet = self.api.starknet.read().await;
 
-        Ok(StarknetResponse::Block(Block {
-            status: *block.status(),
-            header: BlockHeader::from(&block),
-            transactions: starknet_types::rpc::transactions::Transactions::Hashes(
-                block.get_transactions().to_owned(),
-            ),
-        }))
+        let block = starknet.get_block(block_id.as_ref()).map_err(|err| match err {
+            Error::NoBlock => ApiError::BlockNotFound,
+            unknown_error => ApiError::StarknetDevnetError(unknown_error),
+        })?;
+
+        // StarknetBlock needs to be mapped to PendingBlock response only in blocks_on_demand mode
+        // and when block_id is pending
+        if starknet.config.blocks_on_demand
```

Comment:
Since here we only have StarknetBlock type we need to check again `blocks_on_demand` mode and I'm not happy about that. Maybe the type system needs to be changed?